### PR TITLE
Set default values to a secure value - Automation #773

### LINKED ISSUE #773: Set default values to a secure value - Automation
All default values should comply with a security baseline, e.g. [NIST 800](https://docs.microsoft.com/en-us/azure/governance/policy/samples/nist-sp-800-53-r5)

The [build-in policies](https://github.com/Azure/azure-policy) of Azure can be used as a reference.

The task would be to scan over each of the following policies and make sure, that the module is per default complying to them.

The following policies are the NIST 800 ones:

\built-in-policies\policyDefinitions\Automation\AutomationAccount_CMK_Audit.json
\built-in-policies\policyDefinitions\Automation\Automation_AuditUnencryptedVars_Audit.json

### PR DESCRIPTION
**Pull request by elanzel**
# Change
- Updated parameters.json with encryption block
- Updated deploy.bicep encryption identity, keySource, keyVaultProperties, keyVersion, keyvaultUri 
- updated readme.md parameters section and added example for encryption
- updated API version in arm\Microsoft.Automation\automationAccounts\.bicep\nested_privateEndpoint.bicep for Pester last API version issue
- updated Dependencies workflow and pipeline with a new purge protected keyvault
- updated dependencies/kayvault with new nopr.parameters.json

Badge: [Automation: AutomationAccounts](https://github.com/Azure/ResourceModules/actions/workflows/ms.automation.automationaccounts.yml)

Close #773

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
         "schedules": {
             "value": [
                 {
-                    "name": "TestSchedule",
+                    "name": "TestSchedule01",
```
Comment (MrMCake) [inline arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json:11]: why the renaming?

File: arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json [hunk 2]
```diff
@@ -30,7 +30,7 @@
         "runbooks": {
             "value": [
                 {
-                    "name": "TestRunbook",
+                    "name": "TestRunbook01",
```
Comment (MrMCake) [inline arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json:33]: why the renaming?

File: arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json [hunk 3]
```diff
@@ -66,18 +66,18 @@
...
                     "value": "\"TestEncryptedValue\"",
-                    "description": "TestEncryptedDescription",
-                    "isEncrypted": true
+                    "description": "TestEncryptedDescription"
                 }
             ]
         },
         "linkedWorkspaceId": {
-            "value": "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/microsoft.operationalinsights/workspaces/adp-<<namePrefix>>-az-law-x-001"
+            "value": "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/microsoft.operationalinsights/workspaces/adp-<<namePrefix>>-az-law-x-002"
```
[the remark this answers is not in the document]
Comment (eriqua) [reply arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json:80]: I'm creating a dedicated LAW for the automation account to be used, I'll post it here later. The `001 `gets linked to another automation account in our environment (ref. issue #965 ) and is thus not possible to be used to create another link

[the remark this answers is not in the document]
Comment (eriqua) [reply arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json:80]: Please replace with 
```suggestion
            "value": "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/microsoft.operationalinsights/workspaces/adp-<<namePrefix>>-az-law-aut-001"
```

File: arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json (hunk 1 above)
Comment (elbatane) [reply arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json:11]: We had conflicts

Comment (elbatane) [reply arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json]: Already changed to to original value

File: arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json (hunk 2 above)
Comment (elbatane) [reply arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json]: Already changed bacl to its original value

File: arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json [hunk 4]
```diff
@@ -41,8 +41,8 @@
         "jobSchedules": {
             "value": [
                 {
-                    "scheduleName": "TestSchedule",
-                    "runbookName": "TestRunbook"
+                    "scheduleName": "TestSchedule01",
```
[the remark this answers is not in the document]
Comment (elbatane) [reply arm/Microsoft.Automation/automationAccounts/.parameters/parameters.json]: Already changed back to its original value

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json [hunk 5]
```diff
@@ -0,0 +1,63 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "name": {
+      "value": "adp-<<namePrefix>>-az-kv-nopr-001"  // keyvault with purge protection enabled, to be used for encryption
```
Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:6]: why not 'nopurge'? too long?

Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:6]: why not 'nopurge'? too long?

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+              "all"
+            ],
+            "certificates": [
+              "all"
+            ]
+          }
+        }
+      ]
+    },
+    "enablePurgeProtection": {
```
Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:59]: Please add the softDeleteRetentionInDays property and set its value to the minimum

Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:59]: Please add the softDeleteRetentionInDays property and set its value to the minimum

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "name": {
+      "value": "adp-<<namePrefix>>-az-kv-nopr-001"  // keyvault with purge protection enabled, to be used for encryption
+    },
+    "accessPolicies": {
+      "value": [
+        {
+          "tenantId": "<<tenantId>>",
+          "objectId": "bab643ad-2988-4560-9f2b-c8282ba345f0", // Backup Management Service needs access to back up the keys and secrets, along with the associated VMs
```
Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:12]: are the access policies the minimum ones? is the backup management needed?

Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:12]: are the access policies the minimum ones? is the backup management needed?

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/parameters.json [hunk 8]
```diff
@@ -28,7 +28,9 @@
           "tenantId": "<<tenantId>>",
           "objectId": "<<msiPrincipalID>>", // The object ID of the deployed MSI. Replaced by the pipeline
           "permissions": {
-            "keys": [],
+            "keys": [
+              "Get"
```
Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/parameters.json:32]: where is this needed?

Comment (eriqua) [inline utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/parameters.json:32]: where is this needed?

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json [hunk 9]
```diff
@@ -0,0 +1,207 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "<<namePrefix>>-wd-aut-x-002"
```
Comment (eriqua) [inline arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:6]:
```suggestion
            "value": "<<namePrefix>>-wd-aut-encr-001"
```
```suggestion
            "value": "<<namePrefix>>-wd-aut-x-002"
```

Comment (eriqua) [inline arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:6]:
```suggestion
            "value": "<<namePrefix>>-wd-aut-encr-001"
```
```suggestion
            "value": "<<namePrefix>>-wd-aut-x-002"
```

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json (hunk 5 above)
Comment (elanzel) [reply utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:6]: with 'nopurge' KV name was too long (27 chars while are 24 allowed)

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json (hunk 6 above)
Comment (elanzel) [reply utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:59]: added 
"softDeleteRetentionInDays" : { 
      "value": 7
    }

Comment (elanzel) [reply utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:50]: added softDeleteRetentionInDays=7 on a new KV "adp-<<namePrefix>>-az-kv-nopr-002" as the property cannot be modified when it is already in place.
Deleted adp-<<namePrefix>>-az-kv-nopr-001.

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json (hunk 7 above)
Comment (elbatane) [reply utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/nopr.parameters.json:12]: Backp management is not needed so removed

File: utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/parameters.json (hunk 8 above)
Comment (elbatane) [reply utilities/pipelines/dependencies/Microsoft.KeyVault/vaults/parameters/parameters.json:32]: No needed, removed

**Comment by MrMCake** [comment]
[ADO ref](https://dev.azure.com/servicescode/infra-as-code-source/_build/results?buildId=230896&view=logs&j=d357320e-1c02-5461-99a0-dbc7eeb90104)

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 10]
```diff
@@ -584,6 +584,11 @@ stages:
               displayName: Default Key Vault
               jobName: default_kv
               customParameterFileTokens: '{"msiPrincipalId":"$(msiPrincipalId)"}'
+            - path: $(dependencyPath)/$(resourceType)/parameters/nopr.parameters.json
+              templateFilePath: $(templateFilePath)
+              displayName: Default Key Vault
```
Comment (MrMCake) [inline .azuredevops/platformPipelines/platform.dependencies.yml:589]:
```suggestion
              displayName: Purge Proteced KV
```

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json [hunk 11]
```diff
@@ -0,0 +1,207 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "<<namePrefix>>-wd-aut-encr-001"
+        },
+        // "schedules": {
```
Comment (MrMCake) [inline arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:8]: Do we need a schedule?

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        "keyName" : {
+                "value" : "keyEncryptionKey"
+        },
+        "keyvaultUri" : {
+                    "value" : "https://adp-carml-az-kv-nopr-002.vault.azure.net/"
+        },
+        "keyVersion" : {
+                    "value" : "9917c14be51d4d93b37218de7d326f60"
+        },
+        // "modules": {
```
Comment (MrMCake) [inline arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:36]: If we don't test the modules, I'd suggest to remove them. same for the other parameters

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json (hunk 11 above)
Comment (elbatane) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json]: Removed and tested  ( they were commented for testing purposes)

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json (hunk 12 above)
Comment (elbatane) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json]: Removed and tested (They were commented for testing purposes)

**Review by MrMCake — APPROVED** [review]

**Pull request merged**